instance/utils.py

```diff
@@ -218,3 +219,44 @@ def choices(cls):
     def names(cls):
         """Return enum as list of string names """
         return list(prop.name for prop in cls)
+
+
+def create_new_deployment(
+        instance,
+        creator=None,
+        deployment_type=None,
+        cancel_pending=False,
+        **kwargs,
+):
+    """
+    Create a new deployment for an existing instance, and start it asynchronously
+    """
+    # pylint: disable=cyclic-import, useless-suppression
+    from instance.models.deployment import DeploymentType
+    from instance.models.openedx_deployment import DeploymentState, OpenEdXDeployment
+    from instance.tasks import start_deployment
+    from userprofile.models import UserProfile
+
+    changes = None
+    if instance.betatestapplication_set.exists():
+        changes = build_instance_config_diff(instance.betatestapplication_set.get())
+
+    deployment_type = deployment_type or DeploymentType.unknown
+    creator_profile = creator and UserProfile.objects.get(user_id=creator)
+    if cancel_pending:
+        deployment = instance.get_latest_deployment()
+        if deployment.status() in (DeploymentState.changes_pending, DeploymentState.provisioning):
+            deployment.cancel_deployment()
+    deployment = OpenEdXDeployment.objects.create(
+        instance_id=instance.ref.id,
+        creator=creator_profile,
+        type=deployment_type,
+        changes=changes,
+    )
+    # We schedule a deployment instead of starting right now, which gives user a few mins to update the instance
+    # without us having to setup and teardown the appserver.
+    start_deployment.schedule(
+        args=(instance.ref.id, deployment.id),
+        kwargs=kwargs,
+        delay=settings.DEPLOYMENT_START_DELAY
```

Comment:
Would it make sense to only use the delay when deploying from the new end-user interface?

I think it could be confusing for admins if they click the "Launch new appserver" button and nothing seems to happen when using the manage.opencraft.com interface (we'll have to update that interface to show deployments rather than appservers at some point).

For deployments triggered by PR watchers and periodic builds the delay probably also isn't neccessary.

Reply:
Good catch - fixed. Adding delay only for console-triggered deployments